Make rollover 0.4.7 compatible with current master

da-kami opened the pull request:
Introduce `ProposeRolloverV2` message on the wire:

- The taker sends `ProposeRolloverV2`
- The maker can handle both `ProposeRolloverV2` and `ProposeRollover`

Introduce `RolloverVersion` to distinguish code paths in `rollover_maker` actor and `accept_rollover_proposal` for the maker.

I opted for having `RolloverVersion` because the decision logic is simple, and we have to store the version in the `rollover_maker` actor because the proposal does not need the decision logic, only after accepting we decide. It might still be somewhat cleaner to duplicate `accept_rollover_proposal` and deprecate one of the functions - but since I needed the `RolloverVersion` anyway I pulled it in there. 
We could also duplicate all code on the way: the `rollover_maker` actor and  `accept_rollover_proposal` in `model::cfd::Cfd` and mark all duplicated code as deprecated. But I found that a bit of an overkill.

da-kami committed:
Make rollover 0.4.7 compatible with current master

Introduce `ProposeRolloverV2` message on the wire:

- The taker sends `ProposeRolloverV2`
- The maker can handle both `ProposeRolloverV2` and `ProposeRollover`

Introduce `RolloverVersion` to distinguish code paths in `rollover_maker` actor and `accept_rollover_proposal` for the maker.

thomaseizinger commented on model/src/rollover.rs:
```diff
@@ -5,6 +5,12 @@ use crate::Price;
+#[derive(Debug, Clone, Copy)]
+pub enum Version {
+    V1_0_4_7,
+    Current,
```
Just make these 1 and 2?

I think this should version the rollover protocol itself, not which version of the daemon this is implemented.

Also "current" is not a version :)

thomaseizinger replied:
This enum would also be a good place to document the difference in the 2 versions of the protocol!

da-kami replied:
I agree with what you are pointing out and will rename to `V1` and `V2` but probably still mention in rustdoc for what version `V1` was for if that's OK. 

I feel "if this grows", i.e. if we at some point have multiple version it would be great to know for what version we introduced them. That could make it easier to monitor what to remove when we know that there are no takers running 0.4.7 anymore.

klochowicz replied:
this sounds like a good idea - in fact, this would be in line with what I've already done for `HelloV2` message (rustdoc comment that `Hello` was used by takers up to 0.4.7).

thomaseizinger reviewed (commented):
Some thoughts, I'll leave you to it :)

luckysori commented on model/src/cfd.rs:
```diff
@@ -16,6 +17,7 @@ use crate::OpeningFee;
 use crate::Percent;
 use crate::Position;
 use crate::Price;
+use crate::RolloverVersion;
```
🔧 I don't think we need this type. It should be enough to use `rollover::Version` in this file.

luckysori commented on model/src/lib.rs:
```diff
@@ -33,6 +33,7 @@ mod rollover;
 pub use cfd::*;
 pub use contract_setup::SetupParams;
 pub use rollover::RolloverParams;
+pub use rollover::Version as RolloverVersion;
```
:+1: I agree with exporting without leaking the internal structure of `model`.

luckysori commented on model/src/cfd.rs:
```diff
@@ -833,7 +836,11 @@ impl Cfd {
+        let hours_to_charge = match version {
+            Version::V1_0_4_7 => 1,
+            Version::Current => self.hours_to_extend_in_rollover()?,
+        };
```
👍 Having discussed the merits of duplicating code vs introducing branching logic I think this is okay. We can always revisit this choice in the future if we notice that we're having to introduce a lot of code like this.

luckysori reviewed (approved):
I approve.

luckysori commented on model/src/rollover.rs:
```diff
@@ -5,6 +5,21 @@ use crate::Price;
+    /// Version one of the rollover protocol
+    ///
+    /// This version cannot handle charging for "missed" rollovers yet, i.e. the hours to charge is
+    /// always set to 1 hour. This version is needed for clients that are <= daemon version
+    /// `0.4.7`.
```
I understand why you're putting this comment here, but it's a bad sign that our model "knows" about daemon versions. Nothing to action, just a thought.

da-kami replied:
I totally agree, I will see to given our versioning ADR a review later!

klochowicz replied:
From an utilitarian POV, I'd say that the model ha also 0.4.7 version :)

da-kami committed:
Backwards compatibility for bug where current funding_fee is not applied to rollover

The bug is fixed in a backwards compatible way but it is still dangerous to use `0.4.7` because it *can* happen in unlikely cases that the parties come up with different fee values.
If that happens they will, unfortunately, won't notice, the rollover will succeed but the database will contain different `funding_fee` values that will cause trouble on all consecutive rollovers.
All consecutive rollovers will then fail and the CFD will eventually be force closed.
